pydeseq2/ds.py

```diff
@@ -634,27 +636,55 @@ def _build_contrast(self, contrast: Optional[List[str]] = None) -> None:
         if contrast is not None:  # Test contrast if provided
             if len(contrast) != 3:
                 raise ValueError("The contrast should contain three strings.")
-            if contrast[0] not in self.dds.design_factors:
+            if contrast[0] not in self.dds.design_factors_list:
                 raise KeyError(
                     f"The contrast variable ('{contrast[0]}') should be one "
                     f"of the design factors."
                 )
             if not (contrast[1] == contrast[2] == ""):
                 # The contrast factor is categorical, so we should check that the tested
                 # and reference levels are valid.
-                if contrast[1] not in self.dds.obs[contrast[0]].values:
+
+                if ":" in contrast[0]:
+                    columns_interacting = contrast[0].split(":")
+
+                    categorical_columns_interacting = [
+                        col
+                        for col in columns_interacting
+                        if not is_numeric_dtype(self.dds.obs[col])
+                    ]
+
+                    potential_values = [
+                        "".join(interacting_values)
+                        for interacting_values in list(
+                            itertools.product(
+                                *[
+                                    self.dds.obs[col].cat.categories.tolist()
+                                    for col in categorical_columns_interacting
+                                ]
+                            )
+                        )
+                    ]
+
+                else:
+                    categorical_columns_interacting = [contrast[0]]
+                    potential_values = self.dds.obs[contrast[0]].values
+
+                if contrast[1] not in potential_values:
                     raise KeyError(
                         f"The tested level ('{contrast[1]}') should correspond to "
-                        f"one of the levels of '{contrast[0]}'"
+                        "one of the levels or product thereof of"
+                        f" '{categorical_columns_interacting}'"
                     )
-                if contrast[2] not in self.dds.obs[contrast[0]].values:
+                if contrast[2] not in potential_values:
                     raise KeyError(
                         f"The reference level ('{contrast[2]}') should correspond to "
-                        f"one of the levels of '{contrast[0]}'"
+                        "one of the levels or products thereof of"
+                        f" '{categorical_columns_interacting}'"
                     )
             self.contrast = contrast
         else:  # Build contrast if None
-            factor = self.dds.design_factors[-1]
+            factor = self.dds.design_factors_list[-1]
```

Comment:
Indeed. I had failing tests due to the fact that `self.dds.design_factors` should be a string representing a formula, whereas we want to have access to individual factors

Reply:
Thanks will have a look !

Reply:
It's weird because my tests were passing...and if I did that this way it had a purpose ofc I can't remember. Note to my future self add more comments.